### DISCUSSION TITLE
Using Postprocessor on blocks

### DISCUSSION THREAD

**Discussion by hokonkim** [idaholab/moose#18363] [Q&A General] [answered]
While I run several simulations, some conditions gave me values, but some didn't.

I tried to figure it out what was the difference between those simulations through input file.

Case 1. Some codes in an input file giving me a value

    [Mesh]
      [./interface_YSZ_CA]
        type = SideSetsBetweenSubdomainsGenerator
        input = meshScale
        primary_block = 'PT_YSZ_C_TET4'
        paired_block = 'PT_YSZ_A_TET4'
        new_boundary = 'interface_CA'
      [../]

      [./interface_YSZ_AC]
        type = SideSetsBetweenSubdomainsGenerator
        input = interface_YSZ_CA
        primary_block = 'PT_YSZ_A_TET4'
        paired_block = 'PT_YSZ_C_TET4'
        new_boundary = 'interface_AC'
      [../]
    []

    [Materials]
      [./CounterElectrode_for_C]
        type  = ParsedMaterial
        block = 'PT_YSZ_C_TET4 PT_TPB_C_TET4'
        f_name = 'eff_pO2_A'
        constant_names        = 'R          T        z     F'        # (J/K/mol), (K),  (# of electron), (C/mol)
        constant_expressions  = '8.3144598  1073.0   4   96485.3329'
        postprocessor_names = 'numerical_phi_YSZ_C'
        function = 'if(numerical_phi_YSZ_C=0, exp(z * F / R / T * (-0.55442)), exp(z * F / R / T * (numerical_phi_YSZ_C)))'
      [../]

      [./CounterElectrode_for_A]
        type  = ParsedMaterial
        block = 'PT_YSZ_A_TET4 PT_TPB_A_TET4'
        f_name = 'eff_pO2_C'
        constant_names        = 'R          T        z     F'        # (J/K/mol), (K),  (# of electron), (C/mol)
        constant_expressions  = '8.3144598  1073.0   4   96485.3329'
        postprocessor_names = 'numerical_phi_YSZ_A'
        function = 'if(numerical_phi_YSZ_A=0, exp(z * F / R / T * (-0.55442)), exp(z * F / R / T * (numerical_phi_YSZ_A)))'
      [../]
    []

    [Postprocessors]
      [./I_YSZ_bottom_Cathode]
        type = SideDiffusiveFluxIntegral
        variable = phi_YSZ_C
        diffusivity = 'sigma_YSZ'
        boundary = 'interface_CA'
        outputs = 'console csv'
      [../]

      [./numerical_phi_YSZ_C]
        type = AverageNodalVariableValue
        boundary = 'interface_CA'
        variable = phi_YSZ_C
        execute_on = 'timestep_begin nonlinear timestep_end'
      [../]

      [./I_YSZ_bottom_Anode]
        type = SideDiffusiveFluxIntegral
        variable = phi_YSZ_A
        diffusivity = 'sigma_YSZ'
        boundary = 'interface_AC' 
        outputs = 'console csv'
      [../]

      [./numerical_phi_YSZ_A]
        type = AverageNodalVariableValue
        boundary = 'interface_AC'
        variable = phi_YSZ_A
        execute_on = 'timestep_begin nonlinear timestep_end'
      [../]
    []

Case 2. Some codes in an input file **not giving** me a value

    [Mesh]
      [./interface_YSZ_CA]
        type = SideSetsBetweenSubdomainsGenerator
        input = meshScale
        primary_block = 'PT_YSZ_C_TET4'
        paired_block = 'PT_YSZ_A_TET4'
        new_boundary = 'interface_CA'
      [../]

      [./interface_YSZ_AC]
        type = SideSetsBetweenSubdomainsGenerator
        input = interface_YSZ_CA
        primary_block = 'PT_YSZ_A_TET4'
        paired_block = 'PT_YSZ_C_TET4'
        new_boundary = 'interface_AC'
      [../]
    []

    [Materials]
      [./CounterElectrode]
        type  = ParsedMaterial
        block = 'PT_YSZ_C_TET4 PT_TPB_C_TET4 PT_YSZ_A_TET4 PT_TPB_A_TET4'
        f_name = 'eff_pO2'
        constant_names        = 'R          T        z     F'        # (J/K/mol), (K),  (# of electron), (C/mol)
        constant_expressions  = '8.3144598  1073.0   4   96485.3329'
        postprocessor_names = 'numerical_phi_YSZ'
        function = 'if(numerical_phi_YSZ=0, exp(z * F / R / T * (-0.55442)), exp(z * F / R / T * (numerical_phi_YSZ)))'
      [../]
    []

    [Postprocessors]
      [./I_YSZ_bottom_Cathode]
        type = SideDiffusiveFluxIntegral
        variable = phi_YSZ_C
        diffusivity = 'sigma_YSZ'
        boundary = 'interface_CA'
        outputs = 'console csv'
      [../]

      [./I_YSZ_bottom_Anode]
        type = SideDiffusiveFluxIntegral
        variable = phi_YSZ_A
        diffusivity = 'sigma_YSZ'
        boundary = 'interface_AC' 
        outputs = 'console csv'
      [../]

      [./numerical_phi_YSZ]
        type = AverageNodalVariableValue
        boundary = 'interface_CA'
        variable = phi_YSZ_C
        execute_on = 'timestep_begin nonlinear timestep_end'
      [../]
    []

Technically, I tried to integrate the CounterElectrode by introducing one postprocessor "numerical_phi_YSZ" at Case 2 and use the number coming from the postprocessor "numerical_phi_YSZ" for calculating CounterElectrode. 

But, I didn't get values for I_YSZ_bottom_Anode at Case 2.

I think a postprocessor is entangled with a variable (phi_YSZ_C).

So, even if I passed the postprocessor value to CounterElectrode applied on "TPT_YSZ_A_TET4 PT_TPB_A_TET4", it would not work.

Those blocks did not have a variable (phi_YSZ_C), but it had a variable (phi_YSZ_A).

In my opinion, if I use still two variables (phi_YSZ_A, phi_YSZ_C), I should use two postprocessors even two functions work the same.

Also, if I want to optimize the Case 1, I could use **one boundary(interface_CA or interface_AC)**.

Am I correct?

**Comment by GiudGiud** [comment, marked as answer]
Hello

What do you mean you did not get a value? Do you mean you got a 0 output?

Is there anything that creates a gradient on either side of the anode/cathode interface? What is the connection between the two variables? Is there an interface kernel?
Do you need to use two variables? Do you expect I_YSZ_bottom_Cathode and I_YSZ_bottom_Anode to be different given that they are on both sides of the same interface (eg is flux supposed to be discontinuous?)

Guillaume
